dcicutils/creds_utils.py

```diff
@@ -0,0 +1,272 @@
+            cls.KEYS_FILE = filename or cls._default_keys_file(for_testing=True)
+            yield
+        finally:
+            cls.KEYS_FILE = old_filename
+
+    @classmethod
+    def register(cls, *, name):
+        def _register_class(key_manager_class):
+            assert issubclass(key_manager_class, KeyManager)
+            if name in _KEY_MANAGERS:
+                raise ValueError(f"A KeyManager named {name!r} has already been defined.")
+            key_manager_class._init_class_variables()
+            key_manager_class._REGISTERED = True
+            _KEY_MANAGERS[name] = cls
+            return key_manager_class
+        return _register_class
+
+    @classmethod
+    def _init_class_variables(cls):
+        class_name = cls.__name__
+        print(f"cls={cls!r}, name={class_name!r}")
+        suffix = "KeyManager"
+        if not class_name.endswith(suffix):
+            raise ValueError(f"The name, {class_name!r}, of a KeyManager class to be registered"
+                             f" does not end in {suffix!r}.")
+        app_name = remove_suffix(suffix=suffix, text=class_name, required=True)
+        if cls.APP_NAME is None:
+            cls.APP_NAME = app_name
+        elif cls.APP_NAME.lower() != app_name.lower():
+            raise ValueError(f"A KeyManager class with APP_NAME = {cls.APP_NAME!r} expects a name"
+                             f" like {cls.APP_NAME}{suffix}.")
+        else:
+            app_name = cls.APP_NAME
+        if cls.APP_TOKEN is None:
+            cls.APP_TOKEN = app_token = app_name.upper()
+        else:
+            app_token = cls.APP_TOKEN
+            if not app_token.isupper():
+                raise ValueError(f"The APP_TOKEN {app_token!r} must be all-uppercase.")
+            elif app_token[:1] != app_name[0].upper():
+                raise ValueError(f"The APP_TOKEN {app_token!r} must have the same first letter as {app_name!r}")
+        app_prefix = app_token + "_"
+        if not cls.KEYS_FILE_VAR:
+            cls.KEYS_FILE_VAR = f"{app_prefix}KEYS_FILE"
+        elif not cls.KEYS_FILE_VAR.startswith(app_prefix):
+            raise ValueError("The {class_name}.KEYS_FILE_VAR must begin with {app_prefix!r}.")
+        if not cls.KEYS_FILE:
+            cls.KEYS_FILE = cls._default_keys_file()
+        # print(f"Defining {class_name} "
+        #       f"with APP_NAME={app_name!r} APP_TOKEN={app_token!r} KEYS_FILE_VAR={cls.KEYS_FILE_VAR!r}.")
+
+    @classmethod
+    def create(cls, name, **kwargs):
+        key_manager_class = _KEY_MANAGERS.get(name)
+        if not key_manager_class:
+            raise ValueError(f"There is no registered KeyManager class named {name!r}.")
+        return key_manager_class(**kwargs)
+
+    # @contextlib.contextmanager
+    # def alternate_keys_file_from_environ(self):
+    #     filename = os.environ.get(self.KEYS_FILE_VAR) or None  # Treats empty string as undefined
+    #     with self.alternate_keys_file(filename=filename):
+    #         yield
```

Comment:
Several functions are commented out, it isn't necessarily clear which are truly intended to be.

Reply:
Ah, yes. SubmitCGAP had these as functions but I think they were not needed, so I simplified the paradigm. Basically, some things want to be class variables but need inits, so what I did to initialize them functionally was to use a decorator that runs after the class definition and knows how to do the initializations. I think it may be that in later Python versions you can write an init class kind of thing, or maybe I've just seen it as something users can write using metaclasses, but I think the decorator solution is adequate and in some ways easier to think about.